math/src/field/f64/mod.rs

```diff
@@ -53,26 +59,95 @@ const ELEMENT_BYTES: usize = core::mem::size_of::<u64>();
 #[derive(Copy, Clone, Debug, Default)]
 pub struct BaseElement(u64);
-
 impl BaseElement {
-    /// Creates a new field element from the provided `value`. If the value is greater than or
-    /// equal to the field modulus, modular reduction is silently performed.
-    pub const fn new(value: u64) -> Self {
-        Self(value % M)
+    /// Creates a new field element from the provided `value`; the value is converted into
+    /// Montgomery representation.
+    pub const fn new(value: u64) -> BaseElement {
+        BaseElement(BaseElement::mont_red((value as u128) * (R2 as u128)))
+    }
+    /// Gets the inner value that might not be canonical
+
+    pub const fn inner(self: &Self) -> u64 {
+        return self.0;
+    }
+
+    /// Montgomery reduction
+    #[inline(always)]
+    pub const fn mont_red(x: u128) -> u64 {
+        // See reference above for a description of the following implementation.
+        let xl = x as u64;
+        let xh = (x >> 64) as u64;
+        let (a, e) = xl.overflowing_add(xl << 32);
+
+        let b = a.wrapping_sub(a >> 32).wrapping_sub(e as u64);
+
+        let (r, c) = xh.overflowing_sub(b);
+        r.wrapping_sub(0u32.wrapping_sub(c as u32) as u64)
+    }
+
+    /// Addition in BaseField
+    #[inline(always)]
+    const fn add(self, rhs: Self) -> Self {
+        // We compute a + b = a - (p - b).
+        let (x1, c1) = self.0.overflowing_sub(M - rhs.0);
+        let adj = 0u32.wrapping_sub(c1 as u32);
+        BaseElement(x1.wrapping_sub(adj as u64))
+    }
+
+    /// Subtraction in BaseField
+    #[inline(always)]
+    const fn sub(self, rhs: Self) -> Self {
+        // See reference above for more details.
+        let (x1, c1) = self.0.overflowing_sub(rhs.0);
+        let adj = 0u32.wrapping_sub(c1 as u32);
+        BaseElement(x1.wrapping_sub(adj as u64))
+    }
+
+    /// Multiplication in BaseField
+    #[inline(always)]
+    const fn mul(self, rhs: Self) -> Self {
+        // If x < p and y < p, then x*y <= (p-1)^2, and is thus in
+        // range of mont_red().
+        BaseElement(BaseElement::mont_red((self.0 as u128) * (rhs.0 as u128)))
+    }
+
+    /// Squaring in BaseField
+    #[inline(always)]
+    pub const fn square(self) -> Self {
+        self.mul(self)
+    }
+
+    /// Multiple squarings in BaseField: return x^(2^n)
+    pub fn msquare(self, n: u32) -> Self {
+        let mut x = self;
+        for _ in 0..n {
+            x = x.square();
+        }
+        x
+    }
+
+    /// Test of equality between two BaseField elements; return value is
+    /// 0xFFFFFFFFFFFFFFFF if the two values are equal, or 0 otherwise.
+    #[inline(always)]
+    pub const fn equals(self, rhs: Self) -> u64 {
+        // Since internal representation is canonical, we can simply
+        // do a xor between the two operands, and then use the same
+        // expression as iszero().
+        let t = self.0 ^ rhs.0;
+        !((((t | t.wrapping_neg()) as i64) >> 63) as u64)
     }
```

Comment:
Rather than having this method here, we should put it into the implementation of `PartialEq` trait.

Also, a question: if the internal representation is canonical, could we don't figure out if two elements are equal by just comparing their internal representations?

Reply:
Indeed, I think this is done this way to achieve a constant-time implementation. If that's not very important for us at the moment I think we should replace it with your suggestion.

Reply:
Ah! I didn't think about this. Let's keep the logic as is then.